com.google.gms:google-services Compatibility Issue - com.google.firebase:firebase-iid is being requested by various other libraries at [[15.1.0,15.1.0]], but resolves to 16.0.0

After updating to the semantinc versioning the OneSignal Gradle plugin fails.
When I remove the OneSignal gradle plugin (v0.10.1) the sync succeed, otherwise it fails with the following error.

### Full Error
```Java
BUILD ERROR
<ij_msg_gr>Gradle import errors<ij_msg_gr><ij_nav><project_path>/app/build.gradle<ij_nav><i><b>project ':app': Unable to build Kotlin project configuration</b><eol>Details: java.lang.reflect.InvocationTargetException: null<eol>Caused by: org.gradle.api.GradleException: The library com.google.firebase:firebase-iid is being requested by various other libraries at [[15.1.0,15.1.0]], but resolves to 16.0.0. Disable the plugin and check your dependencies tree using ./gradlew :app:dependencies.</i>
The library com.google.firebase:firebase-iid is being requested by various other libraries at [[15.1.0,15.1.0]], but resolves to 16.0.0. Disable the plugin and check your dependencies tree using ./gradlew :app:dependencies.
```

### Dependency Tree
The firebase com.google.firebase:firebase-iid package is referenced from the dependencies below.

```
+--- com.google.firebase:firebase-core:16.0.0
|    +--- com.google.firebase:firebase-analytics:16.0.0
|    |    +--- com.google.android.gms:play-services-basement:15.0.1 (*)
|    |    +--- com.google.android.gms:play-services-measurement-base:[15.0.4] -> 15.0.4
|    |    +--- com.google.android.gms:play-services-stats:15.0.1
|    |    |    \--- com.google.android.gms:play-services-basement:[15.0.1] -> 15.0.1 (*)
|    |    +--- com.google.firebase:firebase-analytics-impl:[16.0.0] -> 16.0.0
|    |    |    +--- com.google.android.gms:play-services-ads-identifier:15.0.1
|    |    |    |    \--- com.google.android.gms:play-services-basement:[15.0.1,16.0.0) -> 15.0.1 (*)
|    |    |    +--- com.google.android.gms:play-services-basement:15.0.1 (*)
|    |    |    +--- com.google.android.gms:play-services-measurement-base:[15.0.4] -> 15.0.4
|    |    |    +--- com.google.android.gms:play-services-stats:15.0.1 (*)
|    |    |    +--- com.google.android.gms:play-services-tasks:15.0.1 (*)
|    |    |    +--- com.google.firebase:firebase-common:16.0.0
|    |    |    |    +--- com.google.android.gms:play-services-basement:15.0.1 (*)
|    |    |    |    \--- com.google.android.gms:play-services-tasks:15.0.1 (*)
|    |    |    \--- com.google.firebase:firebase-iid:16.0.0
|    |    |         +--- com.google.android.gms:play-services-basement:15.0.1 (*)
|    |    |         +--- com.google.android.gms:play-services-stats:15.0.1 (*)
|    |    |         +--- com.google.android.gms:play-services-tasks:15.0.1 (*)
|    |    |         +--- com.google.firebase:firebase-common:16.0.0 (*)
|    |    |         \--- com.google.firebase:firebase-iid-interop:16.0.0
|    |    |              +--- com.google.android.gms:play-services-base:15.0.1 (*)
|    |    |              \--- com.google.android.gms:play-services-basement:15.0.1 (*)
|    |    \--- com.google.firebase:firebase-common:16.0.0 (*)
|    \--- com.google.firebase:firebase-measurement-connector-impl:16.0.0
|         +--- com.google.android.gms:play-services-basement:15.0.1 (*)
|         +--- com.google.android.gms:play-services-measurement-base:[15.0.4] -> 15.0.4
|         +--- com.google.firebase:firebase-analytics:[16.0.0] -> 16.0.0 (*)
|         +--- com.google.firebase:firebase-analytics-impl:[16.0.0] -> 16.0.0 (*)
|         +--- com.google.firebase:firebase-common:16.0.0 (*)
|         \--- com.google.firebase:firebase-measurement-connector:16.0.0
|              \--- com.google.android.gms:play-services-basement:15.0.1 (*)

And

--- com.onesignal:OneSignal:3.9.1
|    +--- com.google.firebase:firebase-messaging:[10.2.1, 12.1.0) -> 12.0.1
|    |    +--- com.google.firebase:firebase-iid:12.0.1 -> 16.0.0 (*)
|    |    +--- com.google.android.gms:play-services-basement:12.0.1 -> 15.0.1 (*)
|    |    +--- com.google.firebase:firebase-common:12.0.1 -> 16.0.0 (*)
|    |    \--- com.google.firebase:firebase-messaging-license:12.0.1
|    +--- com.android.support:support-v4:[26.0.0, 27.2.0) -> 27.1.1 (*)
|    +--- com.android.support:customtabs:[26.0.0, 27.2.0) -> 27.1.1 (*)
|    +--- com.google.android.gms:play-services-location:[10.2.1, 12.1.0) -> 12.0.1
|    |    +--- com.google.android.gms:play-services-base:12.0.1 -> 15.0.1 (*)
|    |    +--- com.google.android.gms:play-services-basement:12.0.1 -> 15.0.1 (*)
|    |    +--- com.google.android.gms:play-services-tasks:12.0.1 -> 15.0.1 (*)
|    |    \--- com.google.android.gms:play-services-location-license:12.0.1
|    \--- com.google.android.gms:play-services-base:[10.2.1, 12.1.0) -> 15.0.1 (*)
```

--- a/build.gradle
+++ b/build.gradle
@@ -32,6 +32,10 @@ dependencies {
     compileOnly 'com.google.gms:google-services:4.0.+'
     testCompileOnly 'com.google.gms:google-services:4.0.+'
 
+    // Google's strict version matcher plugin
+    compileOnly 'com.google.android.gms:strict-version-matcher-plugin:1.0.+'
+    testCompileOnly 'com.google.android.gms:strict-version-matcher-plugin:1.0.+'
+
     testCompile gradleTestKit()
     testCompile 'junit:junit:4.12'
     testCompile('org.spockframework:spock-core:1.0-groovy-2.4') {

--- a/gradle/dependency-locks/compileClasspath.lockfile
+++ b/gradle/dependency-locks/compileClasspath.lockfile
@@ -26,6 +26,7 @@ com.android.tools:dvlib:26.1.2
 com.android.tools:repository:26.1.2
 com.android.tools:sdk-common:26.1.2
 com.android.tools:sdklib:26.1.2
+com.google.android.gms:strict-version-matcher-plugin:1.0.2
 com.google.auto.value:auto-value:1.5.2
 com.google.code.findbugs:jsr305:1.3.9
 com.google.code.gson:gson:2.7

--- a/gradle/dependency-locks/compileOnly.lockfile
+++ b/gradle/dependency-locks/compileOnly.lockfile
@@ -26,6 +26,7 @@ com.android.tools:dvlib:26.1.2
 com.android.tools:repository:26.1.2
 com.android.tools:sdk-common:26.1.2
 com.android.tools:sdklib:26.1.2
+com.google.android.gms:strict-version-matcher-plugin:1.0.2
 com.google.auto.value:auto-value:1.5.2
 com.google.code.findbugs:jsr305:1.3.9
 com.google.code.gson:gson:2.7

--- a/gradle/dependency-locks/testCompileClasspath.lockfile
+++ b/gradle/dependency-locks/testCompileClasspath.lockfile
@@ -26,6 +26,7 @@ com.android.tools:dvlib:26.1.2
 com.android.tools:repository:26.1.2
 com.android.tools:sdk-common:26.1.2
 com.android.tools:sdklib:26.1.2
+com.google.android.gms:strict-version-matcher-plugin:1.0.2
 com.google.auto.value:auto-value:1.5.2
 com.google.code.findbugs:jsr305:1.3.9
 com.google.code.gson:gson:2.7

--- a/gradle/dependency-locks/testCompileOnly.lockfile
+++ b/gradle/dependency-locks/testCompileOnly.lockfile
@@ -26,6 +26,7 @@ com.android.tools:dvlib:26.1.2
 com.android.tools:repository:26.1.2
 com.android.tools:sdk-common:26.1.2
 com.android.tools:sdklib:26.1.2
+com.google.android.gms:strict-version-matcher-plugin:1.0.2
 com.google.auto.value:auto-value:1.5.2
 com.google.code.findbugs:jsr305:1.3.9
 com.google.code.gson:gson:2.7

--- a/src/main/groovy/com/onesignal/androidsdk/GradleProjectPlugin.groovy
+++ b/src/main/groovy/com/onesignal/androidsdk/GradleProjectPlugin.groovy
@@ -141,10 +141,21 @@ class GradleProjectPlugin implements Plugin<Project> {
         moduleVersionOverrides = [:]
         moduleCopied = [:]
 
+        disableGMSVersionChecks()
         detectProjectState()
 
         resolutionHooksForAndroidPluginV3()
         resolutionHooksForAndroidPluginV2()
+    }
+
+    // At fundamental level this OneSignal plugin and the gms version checks are solving the same problem
+    // Disabling the version check part of the gms plugin with their flag
+    static void disableGMSVersionChecks() {
+        project.afterEvaluate {
+            def googleServices = project.extensions.findByName('googleServices')
+            if (googleServices)
+                googleServices.config.disableVersionCheck = true
+        }
     }
 
     // Get the AGP plugin instance if it has been applied already

--- a/src/test/groovy/com/onesignal/androidsdk/MainTest.groovy
+++ b/src/test/groovy/com/onesignal/androidsdk/MainTest.groovy
@@ -578,7 +578,7 @@ class MainTest extends Specification {
         when:
         def results = runGradleProject([
             skipGradleVersion: GRADLE_OLDEST_VERSION,
-            compileLines : compileLines,
+            compileLines : compileLines
         ])
 
         then:
@@ -590,6 +590,26 @@ class MainTest extends Specification {
             assert it.value.contains('com.google.firebase:firebase-auth:15.1.0\n')
         }
     }
+
+
+    def 'Compatible with google-services Gradle Plugin'() {
+        def compileLines = """\
+            compile 'com.google.android.gms:play-services-base:15.0.0'
+            compile 'com.google.firebase:firebase-auth:15.1.0'
+        """
+
+        when:
+        def results = runGradleProject([
+            skipGradleVersion: GRADLE_OLDEST_VERSION,
+            buildscriptDependencies: "classpath 'com.google.gms:google-services:4.0.1'",
+            applyPlugins: "apply plugin: 'com.google.gms.google-services'",
+            compileLines : compileLines
+        ])
+
+        then:
+        assert results // Asserting existence and contains 1+ entries
+    }
+
 
     def 'firebase 15 - keep mixed minor versions'() {
         def compileLines = """\

